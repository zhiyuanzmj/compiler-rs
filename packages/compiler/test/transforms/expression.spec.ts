import { NodeTypes } from '@vue/compiler-dom'
import { describe, expect, test } from 'vitest'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformText,
  transformVOnce,
  type IfIRNode,
} from '../../src'
import { makeCompile } from './_utils'

const compileWithVIf = makeCompile({
  nodeTransforms: [
    transformVOnce,
    transformElement,
    transformText,
    transformChildren,
  ],
})

describe('compiler: expression', () => {
  test('conditional expression', () => {
    const { code, helpers, ir } = compileWithVIf(
      `<>{ok? <span>{msg}</span> : fail ? <div>fail</div>  : null }</>`,
    )

    expect(code).toMatchSnapshot()

    expect(helpers).contains('createIf')

    expect(ir.templates).toEqual(['<span> </span>', '<div>fail</div>'])
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false,
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }],
        },
      },
    })
    expect(ir.block.returns).toEqual([0])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 0 }],
    })

    expect(ir.block.effect).toEqual([])
    expect((op as IfIRNode).positive.effect).lengthOf(0)

    expect(code).matchSnapshot()
  })
  test('logical expression', () => {
    const { code, helpers, ir } = compileWithVIf(
      `<>{ok && <div>{msg}</div>}</>`,
    )

    expect(helpers).contains('createIf')

    expect(ir.templates).toEqual(['<div> </div>'])
    const op = ir.block.dynamic.children[0].operation
    expect(op).toMatchObject({
      type: IRNodeTypes.IF,
      id: 0,
      condition: {
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: 'ok',
        isStatic: false,
      },
      positive: {
        type: IRNodeTypes.BLOCK,
        dynamic: {
          children: [{ template: 0 }],
        },
      },
    })
    expect(ir.block.returns).toEqual([0])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 0 }],
    })

    expect(ir.block.effect).toEqual([])
    expect((op as IfIRNode).positive.effect).lengthOf(0)
    expect(code).toMatchSnapshot()
  })
  test('conditional expression with v-once', () => {
    const { code, helpers, ir } = compileWithVIf(
      `<div>{ok? <span>{msg}</span> : <div>fail</div> }</div>`,
    )
    expect(code).toMatchSnapshot()

    expect(helpers).contains('createIf')
    expect(ir.templates).toEqual([
      '<span> </span>',
      '<div>fail</div>',
      '<div></div>',
    ])
    expect(ir.block.returns).toEqual([5])
    expect(ir.block.dynamic).toMatchObject({
      children: [{ id: 5 }],
    })
  })
})
