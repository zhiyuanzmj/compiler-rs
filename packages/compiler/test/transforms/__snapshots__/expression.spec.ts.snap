// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: expression > conditional expression 1`] = `
"
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (msg))
    return n2
  }, () => {
    const n4 = _createIf(() => (fail), () => {
      const n6 = t1()
      return n6
    }, () => {
      const n8 = _createNodes(null)
      return n8
    })
    return n4
  })
  return n0
"
`;

exports[`compiler: expression > conditional expression 2`] = `
"
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (msg))
    return n2
  }, () => {
    const n4 = _createIf(() => (fail), () => {
      const n6 = t1()
      return n6
    }, () => {
      const n8 = _createNodes(null)
      return n8
    })
    return n4
  })
  return n0
"
`;

exports[`compiler: expression > conditional expression with v-once 1`] = `
"
  const n5 = t2()
  _setInsertionState(n5)
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (msg))
    return n2
  }, () => {
    const n4 = t1()
    return n4
  })
  return n5
"
`;

exports[`compiler: expression > logical expression 1`] = `
"
  const n0 = _createIf(() => (ok), () => {
    const n2 = t0()
    const x2 = _child(n2)
    _setNodes(x2, () => (msg))
    return n2
  }, () => {
    const n4 = _createNodes(() => (ok))
    return n4
  })
  return n0
"
`;
